feat(db): allow external sqlite blobs

alexdachin opened the pull request:
Currently Trilium stores the contents of a File note type inside the SQLite database. While this is extremely nice for simplicity and portability, it can become problematic once you start uploading many reference files (for example some large quality images, large PDFs, videos etc). While in theory SQLite supports huge databases, storing many files in a single database:
- makes the database file harder to backup
- increases database lock times
- increases the database corruption risk slightly

I propose **storing big blobs in the filesystem** and storing a reference to them in the database. The files are stored in the data directory to begin with for simplicity, but it also opens the gate for storing them in other places in the future (like s3 compatible storage systems). This keeps the SQLite database small and performant, enables faster incremental backups (like rsync), reduces the risk of database corruption and still allows Trilium to manage things like note hierarchy with cloning.

In order to not disturb the workflow of other people, this feature is disabled by default (so even large blobs are still stored internally in the database). To enable it you need to set `TRILIUM_EXTERNAL_BLOB_STORAGE` environment variable.

The threshold of storing notes externally vs internally is 100kb by default but can be changed with `TRILIUM_EXTERNAL_BLOB_THRESHOLD` environment variable. This way small blobs are read from the database more efficiently compared to storing all blobs in the filesystem.

Closes https://github.com/TriliumNext/Trilium/issues/6546

alexdachin commented on apps/server/src/becca/entities/abstract_becca_entity.ts:
```diff
@@ -195,6 +198,21 @@ abstract class AbstractBeccaEntity<T extends AbstractBeccaEntity<T>> {
+                blobStorageService.deleteExternal(filePath);
+            }
+        } catch (error) {
+            // contentLocation column might not be present when applying older migrations
```
I tried inlining some of these AbstractBeccaEntity in the 0233__migrate_geo_map_to_collection.ts migration, but some things are event based and the subscribers are using AbstractBeccaEntity as well.

The alternative would be to add a warning that people have to upgrade to the the previous Trilium version first and wait for migrations to complete, but I wanted to avoid that

adoriandoran commented:
Once the data directory is relocated, the external BLOB paths break, since they are relative to Trilium’s directory. I suggest storing these paths in the database as relative to the “external-blobs” directory instead.

alexdachin commented:
I see, thank you all for the feedback! I'm away for a few days, but I'll have a look as soon as I get back

alexdachin commented:
Thank you again for checking this out! I found out what the issue was with the json file and fixed it. Good points regarding case insensitive file systems, I switched to a random uuid instead and stored the relative path. Do you mind checking it one more time? 🙏

eliandoran reviewed (changes requested):
There are two bugs:

The first is a critical one (data loss):

1. Import the Trace.json file. It will get stored as an external blob.
2. Replace the entire text with a few characters to delete the external blob.
3. Paste the Trace.json back in (or any big text to trigger saving).
4. Restart the server.
5. The server will restart with something like:
   ```
   Consistency issue fixed: Note '5zzWKjTnXz82' content was set to '' since it was null even though it is not deleted
   Consistency issue fixed: Note 'i2EbsDSQkdeb' content was set to '{}' since it was null even though it is not deleted
   ```
6. Look at the Trace.json note, the content will be replaced with `{}`.

The second one involves uploading a video. Regardless of the video size, it will not be saved as an external blob.

In addition, please address the topics from the bot.

alexdachin commented on apps/server/src/services/blob-storage.ts:
```diff
@@ -0,0 +1,102 @@
+                fs.mkdirSync(dir, { recursive: true, mode: 0o700 });
+            }
+
+            fs.writeFileSync(absolutePath, content, { mode: 0o600 });
```
Good point that async would be ideal. Currently the whole app is using better-sqlite3, which is synchronous. Before this change, blobs were stored in sqlite and read/written with blocking calls, so this is maintaining that same pattern just with filesystem storage instead.

Making this async would mean refactoring the call chain (and potentially switching to an async sqlite library), but I figured it shouldn't be in scope for this feature, what do you think?

dedmen commented on apps/server/src/services/erase.ts:
```diff
@@ -92,26 +93,34 @@ function eraseRevisions(revisionIdsToErase: string[]) {
 
-    sql.executeMany(/*sql*/`DELETE FROM blobs WHERE blobId IN (???)`, unusedBlobIds);
+    // Clean up external blob files before deleting from database
+    unusedBlobRows.forEach(row => blobStorageService.deleteExternal(row));
```
Just a note.
If just one of the many files fails to delete (I know it's unlikely currently, might become less unlikely if other storage protocols are added later), all files will be left inside the database.
Should not be an issue as they are not referenced anywhere anyway and just lay around unused.

But might be nice to, if _some_ fail to delete, at least remove the ones from database that successfully deleted.

Also if you have one file that always fails, that means no blobs DB entries are ever deleted, even if all others would delete fine.

One case where a file might fail to delete, is if it is currently open in a external application (which shouldn't happen tbh).
Which brings the other idea, currently we export files into a temp file to open them, we don't need to do that if we already have the standalone file in the external-blobs folder, but that is only something to potentially consider later, not now.
